app/controllers/orders_controller.rb

```diff
@@ -48,28 +80,89 @@ def new
   def update
     @order = Order.find(session[:order_id]) #session is persistent, the cookie has the session information.
     @order.update(order_param[:order]) # when you get a request, here is my information for right now.
+
     reduce_inventory(@order)
-    if @order.save
+
+    zip = params["order"]["zip"]
+    city = params["order"]["city"]
+    state = params["order"]["state"]
+    # message = ""
+    message = valid_location(zip, city, state)
+
+
+    if @order.save && message == ""
       redirect_to order_path
     else
+      flash.now[:message] = message if !message.empty?
       render :edit
     end
   end
 
+
+  def valid_location(zip, city, state)
+    zips = valid_zip(zip)
+    message = ""
+
+    if !zips.nil?
+      zip_city = zip.to_region(:city => true)
+      zip_state = zip.to_region(:state => true)
+
+      valid_city = city.downcase.strip != zip_city.downcase
+      valid_state = state.upcase != zip_state
+      message+=" **City does not match zip code. Did you mean: #{zip_city}? " if valid_city
+      message += " **State does not match zip code and or 2 letter abbreviation. Did you mean: #{zip_state}? "if valid_state
+      # flash[:success] = message
+    elsif zips.nil?
+      message += "Invalid Zip Code!"
+      # flash[:success] = message
+    end
+    # raise
+    return message
+  end
+
+  def valid_zip(zip)
+    if zip.length == 5
+      return zip.to_region
+    else
+      return nil
+    end
+  end
+
   def confirmation
     @order = Order.find(params[:id])
     @order.update(completed_time: Time.now, completion_status: "paid")
     @order.save
+
+
+    # @order_items = OrderItem.where(order_id: session[:order_id])
+    # session[:order_id] += 1
+    # destroy all cart items
+    # raise
+  end
+
+  def shipping
+    @order = Order.find(params[:id])
+    hash = JSON.parse(params["rate"])
+    rate = hash.values.first.to_i/100.0
+    type= hash.keys.first
```

Comment:
The use of first here seems brittle to me. Right now, this might work because the JSON in the rate params happens to return the rate and type in the first spot when the keys and values are listed. What will happen if the order of the JSON keys and values changes? Especially if this data comes from an external API, it is not safe to assume that the data you want will always be in the first spot of the array.

How could this be refactored to access the data in a more reliable way?